subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleResolveState.java

```diff
@@ -409,10 +423,34 @@ public void maybeUpdateSelection() {
         if (selected == null) {
             select(newSelected);
         } else if (newSelected != selected) {
+            if (++selectionChangedCounter > MAX_SELECTION_CHANGE) {
+                // Let's ignore modules that are changing selection way too much, by keeping the highest version
+                if (maybeSkipSelectionChange(newSelected)) {
+                    return;
+                }
+            }
             changeSelection(newSelected);
         }
     }
 
+    private boolean maybeSkipSelectionChange(ComponentState newSelected) {
+        if (selectionChangedCounter == MAX_SELECTION_CHANGE + 1) {
+            LOGGER.warn("Version computation for module '{}' changed selection too many times.\n" +
```

Comment:
Suggestion:

The dependency resolution engine wasn't able to find a version of module {} which satisfied all requirements because the graph wasn't stable enough. The highest version was selected in order to stabilize selection. Features available in a stable graph like version alignment are not guaranteed in this case.